File: PSO_Implementation/Traditional_PSO.m

%% Traditional Path Swarm Optimization
% Pavel Berezovsky
% Meng Project
 
%% PSO 
function Global_Optimal_Solution = Traditional_PSO(Func,Population,number_of_variables,Iterations)
    % Upper and lower bounds of the function chosen
    [Lower_Bound, Upper_Bound, ~] = cost_function(zeros(1,number_of_variables),number_of_variables,Func);
 
    % Setting Upper and Lower Bounds for position and velocity vector
    Bounds_p = [Lower_Bound Upper_Bound];
    Bounds_v = [-(Bounds_p(2)-Bounds_p(1)) Bounds_p(2)-Bounds_p(1)];
    Global.best_cost = inf; % The initial best cost
    Global.best_position = inf; % Initial best position
 
    for i = 1:Population
        % Setting initial position of each particle within the bounds of Func
        Particle(i).position = Bounds_p(1) + 2*Bounds_p(2)*rand(1,number_of_variables);
        % Best position of each particle is its current initial position
        Particle(i).best_position = Particle(i).position; 
        % Obtaining best cost of each particle
        [~,~,Particle(i).best_cost]= cost_function(Particle(i).position,number_of_variables,Func); 
        % If best cost of each particle is less than the global cost set
        % global cost
        if Particle(i).best_cost < Global.best_cost
            Global.best_cost = Particle(i).best_cost;
            Global.best_position = Particle(i).best_position;
        end
        % Initial particle velocity is 0
        Particle(i).velocity = zeros(1,number_of_variables);
    end
 
    % PSO LOOP
    % iteration count
    count = 0;
    % Inertial coefficient (memory)
    w = .8; 
    % Coefficients 
    c1 = 2; c2 = 2;
 
    while (count < Iterations)
        for i = 1:Population
            % Random elements added to generate velocity
            r1 = rand;
            r2 = rand;
            % Equation of particle velocity
            Particle(i).velocity = w*Particle(i).velocity...
             +c1*r1*(Particle(i).best_position-Particle(i).position)...
             +c2*r2*(Global.best_position-Particle(i).position);
            % If particle velocity exceeds bounds it is set to bounds
            U_Bounds_Check = find(Particle(i).velocity > Bounds_v(2));
            D_Bounds_Check = find(Particle(i).velocity < Bounds_v(1));
            Particle(i).velocity(U_Bounds_Check) = Bounds_v(2);
            Particle(i).velocity(D_Bounds_Check) = Bounds_v(1);
            % Equation for particle position
            Particle(i).position = Particle(i).position+Particle(i).velocity;
            % If particle position exceeds bounds it is set to bounds
            U_Bounds_Check = find(Particle(i).position > Bounds_p(2));
            D_Bounds_Check = find(Particle(i).position < Bounds_p(1));
            Particle(i).position(U_Bounds_Check) = Bounds_p(2);
            Particle(i).position(D_Bounds_Check) = Bounds_p(1);
         
            % Finding temporary cost of the particle
            [~,~,Temp_cost] = cost_function(Particle(i).position,number_of_variables,Func);
            % If temporary cost is less than particle best cost
            if Temp_cost < Particle(i).best_cost
                Particle(i).best_cost = Temp_cost;
                Particle(i).best_position = Particle(i).position;
            end
            % If global best cost is greater than particle best cost 
            if Particle(i).best_cost < Global.best_cost
                Global.best_cost = Particle(i).best_cost;
                Global.best_position = Particle(i).best_position;
            end 
        end
        % Tracking global cost
        Total(count+1) = Global.best_cost;
        % Incrimenting itterations 
        count = count+1;
        
    end
    Global_Optimal_Solution = Global.best_cost;
end
 

File: PSO_Implementation/PythonImpl/Traditional_PSO.py
"""
Traditional Path Swarm Optimization
Pavel Berezovsky
Meng Project

"""

import numpy as np
import random
from cost_function import cost_function

class Particle():
    def __init__(self, pos, number_of_variables):
        self.position = pos
        self.best_position = pos
        self.best_cost = 0
        self.velocity = np.zeros(number_of_variables)

# PSO
def Traditional_PSO(Func, Population, number_of_variables, Iterations):

    # Upper and lower bounds of the function chosen
    Lower_Bound, Upper_Bound, cost = cost_function(np.zeros(number_of_variables), number_of_variables, Func)
 
    # Setting Upper and Lower Bounds for position and velocity vector
    Bounds_p = np.array([Lower_Bound, Upper_Bound])
    Bounds_v = np.array([-(Bounds_p[1]-Bounds_p[0]), Bounds_p[1]-Bounds_p[0]])
    Global_best_cost = float('inf')        # The initial best cost
    Global_best_position = float('inf')    # Initial best position
 
    Particles = []
    random_vec = np.array([random.random() for i in range(number_of_variables)])

    for i in range(0, Population):
        # Setting initial position of each particle within the bounds of Func
        Particles.append(Particle(Bounds_p[0] + 2 *Bounds_p[1] * random_vec, number_of_variables))
        # Best position of each particle is its current initial position

        # Obtaining best cost of each particle
        discard1, discard2, Particles[i].best_cost = cost_function(Particles[i].position, number_of_variables, Func)

        # If best cost of each particle is less than the global cost set
        # global cost
        if Particles[i].best_cost < Global_best_cost:
            Global_best_cost = Particles[i].best_cost
            Global_best_position = Particles[i].best_position
 
    # PSO LOOP
    # iteration count
    count = 0
    # Inertial coefficient (memory)
    w = .8
    # Coefficients 
    c1 = 2
    c2 = 2
 
    Total = [0 for i in range(0, Iterations)]

    while (count < Iterations):
        for i in range(0, Population):
            # Random elements added to generate velocity
            r1 = random.random()
            r2 = random.random()

            # Equation of particle velocity
            Particles[i].velocity = w*Particles[i].velocity \
                + c1 * r1 * (Particles[i].best_position - Particles[i].position) \
                + c2 * r2 * (Global_best_position - Particles[i].position)

            # If particle velocity exceeds bounds it is set to bounds
            for j in range(0, number_of_variables):
                if Particles[i].velocity[j] > Bounds_v[1]:
                   Particles[i].velocity[j] = Bounds_v[1]
                elif Particles[i].velocity[j] < Bounds_v[0]:
                   Particles[i].velocity[j] = Bounds_v[0]
 
            # Equation for particle position
            Particles[i].position = Particles[i].position + Particles[i].velocity
            
            # If particle position exceeds bounds it is set to bounds
            for j in range(0, number_of_variables):
                if Particles[i].position[j] > Bounds_p[1]:
                   Particles[i].position[j] = Bounds_p[1]
                elif Particles[i].position[j] < Bounds_p[0]:
                   Particles[i].position[j] = Bounds_p[0]

            # Finding temporary cost of the particle
            discard1, discard2, Temp_cost = cost_function(Particles[i].position, number_of_variables, Func)
            
            # If temporary cost is less than particle best cost
            if Temp_cost < Particles[i].best_cost:
                Particles[i].best_cost = Temp_cost
                Particles[i].best_position = Particles[i].position

            # If global best cost is greater than particle best cost 
            if Particles[i].best_cost < Global_best_cost:
                Global_best_cost = Particles[i].best_cost
                Global_best_position = Particles[i].best_position
        
        # Tracking global cost
        Total[count] = Global_best_cost
        
        # Incrementing iterations 
        count = count + 1
        
    return Global_best_cost
 
